### DISCUSSION TITLE
CUDA - Noob question

### DISCUSSION THREAD

**Discussion by PxPete** [emgucv/emgucv#719] [Q&A]
Hi, I am sure I am doing something stupid and this really is my last resort, but I have now run out of ideas...

I have been evaluating Emgu & Open CV for a few weeks now and love it! I am now trying to evaluation using CUDA to speed up object detection (working fine on the CPU just slow) so I have followed the instructions and downloaded the "-windesktop_x64-cuda-4.5.5..." self extracting exe and followed the instructions on how to use that in my test winform app. However, when I call CudaInvoke.GetCudaEnabledDeviceCount() I get an exception: "initialization error", "Gpu API call", "-217". I have, of course, googled for days to find something that would solve this for me and no joy - hence this message.

OclInvoke.GetPlatformsInfo() returns a collection that has my Nvidia card in it and I have made sure all my drivers are up to date.

I know this is me, but if anyone could tell me what I am missing I would really appreciate it.

Thanks
Pete

**Comment by emgucv** [comment]
If you download the GPU package from the 4.5.5 release [libemgucv-windesktop_x64-cuda-4.5.5.4823.zip.selfextract.exe](https://github.com/emgucv/emgucv/releases/download/4.5.5/libemgucv-windesktop_x64-cuda-4.5.5.4823.zip.selfextract.exe)

Open the demo VS solution under the "Solution\Windows.Desktop\Emgu.CV.Example.sln".

Can you run the BuildInfo.NetCore.Console project and shared the output here?

**Reply by PxPete** [reply]
Hi, thanks for getting back to me so quickly. Here is the output...


General configuration for OpenCV 4.5.5 =====================================
  Version control:               4.5.5-237-gaa8b5d2f70

  Extra modules:
    Location (extra):            D:/bb/cv_x64_cuda/build/build_x64/../opencv_contrib/modules
    Version control (extra):     4.5.5-62-g7cb6b4d5

  Platform:
    Timestamp:                   2022-01-04T22:06:54Z
    Host:                        Windows 10.0.22000 AMD64
    CMake:                       3.21.3
    CMake generator:             Visual Studio 16 2019
    CMake build tool:            C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/MSBuild/Current/Bin/MSBuild.exe
    MSVC:                        1929
    Configuration:               Debug Release MinSizeRel RelWithDebInfo

  CPU/HW features:
    Baseline:                    SSE SSE2 SSE3
      requested:                 SSE3

  C/C++:
    Built as dynamic libs?:      YES
    C++ standard:                11
    C++ Compiler:                C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe  (ver 19.29.30138.0)
    C++ flags (Release):         /DWIN32 /D_WINDOWS /W4 /GR  /D _CRT_SECURE_NO_DEPRECATE /D _CRT_NONSTDC_NO_DEPRECATE /D _SCL_SECURE_NO_WARNINGS /Gy /bigobj /Oi  /fp:precise     /EHa /wd4127 /wd4251 /wd4324 /wd4275 /wd4512 /wd4589 /MP  /MD /O2 /Ob2 /DNDEBUG
    C++ flags (Debug):           /DWIN32 /D_WINDOWS /W4 /GR  /D _CRT_SECURE_NO_DEPRECATE /D _CRT_NONSTDC_NO_DEPRECATE /D _SCL_SECURE_NO_WARNINGS /Gy /bigobj /Oi  /fp:precise     /EHa /wd4127 /wd4251 /wd4324 /wd4275 /wd4512 /wd4589 /MP  /MDd /Zi /Ob0 /Od /RTC1
    C Compiler:                  C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe
    C flags (Release):           /DWIN32 /D_WINDOWS /W3  /D _CRT_SECURE_NO_DEPRECATE /D _CRT_NONSTDC_NO_DEPRECATE /D _SCL_SECURE_NO_WARNINGS /Gy /bigobj /Oi  /fp:precise     /MP   /MD /O2 /Ob2 /DNDEBUG
    C flags (Debug):             /DWIN32 /D_WINDOWS /W3  /D _CRT_SECURE_NO_DEPRECATE /D _CRT_NONSTDC_NO_DEPRECATE /D _SCL_SECURE_NO_WARNINGS /Gy /bigobj /Oi  /fp:precise     /MP /MDd /Zi /Ob0 /Od /RTC1
    Linker flags (Release):      /machine:x64  /INCREMENTAL:NO
    Linker flags (Debug):        /machine:x64  /debug /INCREMENTAL
    ccache:                      NO
    Precompiled headers:         NO
    Extra dependencies:          cudart_static.lib nppc.lib nppial.lib nppicc.lib nppidei.lib nppif.lib nppig.lib nppim.lib nppist.lib nppisu.lib nppitc.lib npps.lib cublas.lib cudnn.lib cufft.lib -LIBPATH:C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.3/lib/x64
    3rdparty dependencies:

  OpenCV modules:
    To be built:                 alphamat aruco barcode bgsegm bioinspired calib3d ccalib core cudaarithm cudabgsegm cudacodec cudafeatures2d cudafilters cudaimgproc cudalegacy cudaobjdetect cudaoptflow cudastereo cudawarping cudev datasets dnn dnn_objdetect dnn_superres dpm face features2d flann freetype fuzzy gapi hdf hfs highgui img_hash imgcodecs imgproc intensity_transform line_descriptor mcc ml objdetect optflow phase_unwrapping photo plot quality rapid reg rgbd saliency shape stereo stitching structured_light superres surface_matching text tracking video videoio videostab viz wechat_qrcode world xfeatures2d ximgproc xobjdetect xphoto
    Disabled:                    python_bindings_generator python_tests
    Disabled by dependency:      -
    Unavailable:                 cvv java julia matlab ovis python2 python3 sfm ts
    Applications:                -
    Documentation:               NO
    Non-free algorithms:         NO

  Windows RT support:            NO

  GUI:
    Win32 UI:                    YES
    VTK support:                 YES (ver 8.2.0)

  Media I/O:
    ZLib:                        build (ver 1.2.11)
    JPEG:                        build-libjpeg-turbo (ver 2.1.2-62)
    WEBP:                        build (ver encoder: 0x020f)
    PNG:                         build (ver 1.6.37)
    TIFF:                        build (ver 42 - 4.2.0)
    JPEG 2000:                   build (ver 2.4.0)
    OpenEXR:                     build (ver 2.3.0)
    HDR:                         YES
    SUNRASTER:                   YES
    PXM:                         YES
    PFM:                         YES

  Video I/O:
    DC1394:                      NO
    FFMPEG:                      YES (prebuilt binaries)
      avcodec:                   YES (58.134.100)
      avformat:                  YES (58.76.100)
      avutil:                    YES (56.70.100)
      swscale:                   YES (5.9.100)
      avresample:                YES (4.0.0)
    GStreamer:                   NO
    DirectShow:                  YES
    Media Foundation:            YES
      DXVA:                      YES

  Parallel framework:            Concurrency

  Trace:                         YES (with Intel ITT)

  Other third-party libraries:
    Eigen:                       YES (ver 3.3.7)
    Custom HAL:                  NO
    Protobuf:                    build (3.19.1)

  NVIDIA CUDA:                   YES (ver 11.3, CUFFT CUBLAS)
    NVIDIA GPU arch:             52 60 61 70 75 80 86
    NVIDIA PTX archs:

  cuDNN:                         YES (ver 8.2.0)

  OpenCL:                        YES (NVD3D11)
    Include path:                D:/bb/cv_x64_cuda/build/opencv/3rdparty/include/opencl/1.2
    Link libraries:              Dynamic load

  Python (for build):            C:/python-virt/python37/python.exe

  Java:
    ant:                         NO
    JNI:                         C:/Program Files/Eclipse Foundation/jdk-8.0.302.8-hotspot/include C:/Program Files/Eclipse Foundation/jdk-8.0.302.8-hotspot/include/win32 C:/Program Files/Eclipse Foundation/jdk-8.0.302.8-hotspot/include
    Java wrappers:               NO
    Java tests:                  NO

  Install to:                    D:/bb/cv_x64_cuda/build/build_x64/install
-----------------------------------------------------------------


[ INFO:0@0.539] global D:\bb\cv_x64_cuda\build\opencv\modules\core\src\ocl.cpp (1175) cv::ocl::haveOpenCL Initialize OpenCL runtime...
[ INFO:0@0.935] global D:\bb\cv_x64_cuda\build\opencv\modules\core\src\ocl.cpp (1181) cv::ocl::haveOpenCL OpenCL: found 3 platforms
[ INFO:0@0.935] global D:\bb\cv_x64_cuda\build\opencv\modules\core\src\ocl.cpp (973) cv::ocl::OpenCLExecutionContext::Impl::getInitializedExecutionContext OpenCL: initializing thread execution context
[ INFO:0@0.936] global D:\bb\cv_x64_cuda\build\opencv\modules\core\src\ocl.cpp (983) cv::ocl::OpenCLExecutionContext::Impl::getInitializedExecutionContext OpenCL: creating new execution context...
[ INFO:0@0.993] global D:\bb\cv_x64_cuda\build\opencv\modules\core\src\ocl.cpp (1001) cv::ocl::OpenCLExecutionContext::Impl::getInitializedExecutionContext OpenCL: device=NVS 310
DNN backends:
OpenCV - OpenCL
OpenCV - OpenCLFp16
OpenCV - Cpu
Cuda - Cuda
Cuda - CudaFp16

Has OpenCL: True
Use OpenCL: True
Platform 0: AMD Accelerated Parallel Processing
   Device 0 : Intel(R) Core(TM) i7-4770 CPU @ 3.40GHz 1.2 (Cpu):Version - OpenCL 1.2 AMD-APP (1800.11); Global memory - 17083613184; Local memory - 32768; Max image size - 8192x8192; DoubleFpConfig: Denorm, InfNan, RoundToNearest, RoundToZero, RoundToInf, Fma
Platform 1: NVIDIA CUDA
   Device 0 (Default): NVS 310 1.1 (Gpu):Version - OpenCL 1.1 CUDA; Global memory - 1073741824; Local memory - 49152; Max image size - 16384x16384; DoubleFpConfig: Denorm, InfNan, RoundToNearest, RoundToZero, RoundToInf, Fma
Platform 2: Intel(R) OpenCL
   Device 0 : Intel(R) Core(TM) i7-4770 CPU @ 3.40GHz 1.2 (Cpu):Version - OpenCL 1.2 (Build 10094); Global memory - 17083613184; Local memory - 32768; Max image size - 16384x16384; DoubleFpConfig: Denorm, InfNan, RoundToNearest, RoundToZero, RoundToInf, Fma



Has Cuda: False


[DEBUG:0@1.184] global D:\bb\cv_x64_cuda\build\opencv\modules\videoio\src\videoio_registry.cpp (197) cv::`anonymous-namespace'::VideoBackendRegistry::VideoBackendRegistry VIDEOIO: Builtin backends(8): FFMPEG(1000); GSTREAMER(990); INTEL_MFX(980); MSMF(970); DSHOW(960); CV_IMAGES(950); CV_MJPEG(940); UEYE(930)
[DEBUG:0@1.184] global D:\bb\cv_x64_cuda\build\opencv\modules\videoio\src\videoio_registry.cpp (221) cv::`anonymous-namespace'::VideoBackendRegistry::VideoBackendRegistry VIDEOIO: Available backends(8): FFMPEG(1000); GSTREAMER(990); INTEL_MFX(980); MSMF(970); DSHOW(960); CV_IMAGES(950); CV_MJPEG(940); UEYE(930)
[ INFO:0@1.185] global D:\bb\cv_x64_cuda\build\opencv\modules\videoio\src\videoio_registry.cpp (223) cv::`anonymous-namespace'::VideoBackendRegistry::VideoBackendRegistry VIDEOIO: Enabled backends(8, sorted by priority): FFMPEG(1000); GSTREAMER(990); INTEL_MFX(980); MSMF(970); DSHOW(960); CV_IMAGES(950); CV_MJPEG(940); UEYE(930)
[ INFO:0@1.361] global D:\bb\cv_x64_cuda\build\opencv\modules\videoio\src\backend_plugin.cpp (383) cv::impl::getPluginCandidates Found 2 plugin(s) for FFMPEG
[ INFO:0@1.361] global D:\bb\cv_x64_cuda\build\opencv\modules\core\src\utils\plugin_loader.impl.hpp (67) cv::plugin::impl::DynamicLib::libraryLoad load D:\EmguCV\Emgu.CV.Example\BuildInfo.NetCore.Console\bin\Debug\net6.0\runtimes\win-x64\native\opencv_videoio_ffmpeg455_64.dll => OK
[ INFO:0@1.362] global D:\bb\cv_x64_cuda\build\opencv\modules\videoio\src\backend_plugin.cpp (50) cv::impl::PluginBackend::initCaptureAPI Found entry: 'opencv_videoio_capture_plugin_init_v1'
[ INFO:0@1.362] global D:\bb\cv_x64_cuda\build\opencv\modules\videoio\src\backend_plugin.cpp (169) cv::impl::PluginBackend::checkCompatibility Video I/O: initialized 'FFmpeg OpenCV Video I/O Capture plugin': built with OpenCV 4.5 (ABI/API = 1/1), current OpenCV version is '4.5.5' (ABI/API = 1/1)
[ INFO:0@1.362] global D:\bb\cv_x64_cuda\build\opencv\modules\videoio\src\backend_plugin.cpp (69) cv::impl::PluginBackend::initCaptureAPI Video I/O: plugin is ready to use 'FFmpeg OpenCV Video I/O Capture plugin'
[ INFO:0@1.363] global D:\bb\cv_x64_cuda\build\opencv\modules\videoio\src\backend_plugin.cpp (84) cv::impl::PluginBackend::initWriterAPI Found entry: 'opencv_videoio_writer_plugin_init_v1'
[ INFO:0@1.364] global D:\bb\cv_x64_cuda\build\opencv\modules\videoio\src\backend_plugin.cpp (169) cv::impl::PluginBackend::checkCompatibility Video I/O: initialized 'FFmpeg OpenCV Video I/O Writer plugin': built with OpenCV 4.5 (ABI/API = 1/1), current OpenCV version is '4.5.5' (ABI/API = 1/1)
[ INFO:0@1.364] global D:\bb\cv_x64_cuda\build\opencv\modules\videoio\src\backend_plugin.cpp (103) cv::impl::PluginBackend::initWriterAPI Video I/O: plugin is ready to use 'FFmpeg OpenCV Video I/O Writer plugin'
[ INFO:0@1.714] global D:\bb\cv_x64_cuda\build\opencv\modules\videoio\src\backend_plugin.cpp (383) cv::impl::getPluginCandidates Found 2 plugin(s) for GSTREAMER
[ INFO:0@1.715] global D:\bb\cv_x64_cuda\build\opencv\modules\core\src\utils\plugin_loader.impl.hpp (67) cv::plugin::impl::DynamicLib::libraryLoad load D:\EmguCV\Emgu.CV.Example\BuildInfo.NetCore.Console\bin\Debug\net6.0\runtimes\win-x64\native\opencv_videoio_gstreamer455_64.dll => FAILED
[ INFO:0@1.718] global D:\bb\cv_x64_cuda\build\opencv\modules\core\src\utils\plugin_loader.impl.hpp (67) cv::plugin::impl::DynamicLib::libraryLoad load opencv_videoio_gstreamer455_64.dll => FAILED
[ INFO:0@1.785] global D:\bb\cv_x64_cuda\build\opencv\modules\videoio\src\cap_msmf.cpp (1020) CvCapture_MSMF::configureHW MSMF: Using D3D11 video acceleration on GPU device: NVIDIA NVS 310
[DEBUG:0@1.787] global D:\bb\cv_x64_cuda\build\opencv\modules\videoio\src\cap_msmf.cpp (1198) CvCapture_MSMF::open Device 0 not found (total 0 devices)
[ INFO:0@1.805] global D:\bb\cv_x64_cuda\build\opencv\modules\videoio\src\backend_static.cpp (17) cv::applyParametersFallback VIDEOIO: Backend 'DSHOW' implementation doesn't support parameters in .open(). Applying 0 properties through .setProperty()
[ INFO:0@1.805] global D:\bb\cv_x64_cuda\build\opencv\modules\videoio\src\backend_plugin.cpp (383) cv::impl::getPluginCandidates Found 2 plugin(s) for UEYE
[ INFO:0@1.805] global D:\bb\cv_x64_cuda\build\opencv\modules\core\src\utils\plugin_loader.impl.hpp (67) cv::plugin::impl::DynamicLib::libraryLoad load D:\EmguCV\Emgu.CV.Example\BuildInfo.NetCore.Console\bin\Debug\net6.0\runtimes\win-x64\native\opencv_videoio_ueye455_64.dll => FAILED
[ INFO:0@1.807] global D:\bb\cv_x64_cuda\build\opencv\modules\core\src\utils\plugin_loader.impl.hpp (67) cv::plugin::impl::DynamicLib::libraryLoad load opencv_videoio_ueye455_64.dll => FAILED
Capture Backends (VideoCapture from device):
1900 - FFMPEG
1800 - GSTREAMER
2300 - INTEL_MFX
1400 - MSMF
700 - DSHOW
2000 - CV_IMAGES
2200 - CV_MJPEG
2500 - UEYE

Stream Backends (VideoCapture from file/Stream):
1900 - FFMPEG
1800 - GSTREAMER
2300 - INTEL_MFX
1400 - MSMF
2000 - CV_IMAGES
2200 - CV_MJPEG

VideoWriter backends:
1900 - FFMPEG
1800 - GSTREAMER
2300 - INTEL_MFX
1400 - MSMF
2000 - CV_IMAGES
2200 - CV_MJPEG

Trying to create H263 video writer with FFMPEG backend...
H263 VideoWriter successfully created with backend: FFMPEG (hw acceleration: None)
Trying to create H264 video writer with MSMF backend...
H264 VideoWriter successfully created with backend: MSMF (hw acceleration: None)
Trying to create H263 video writer with default backend...
H263 VideoWriter successfully created with default backend: FFMPEG (hw acceleration: None)


VideoCapture device failed to opened.

Runtime info:
OS Architecture: X64
Framework Description: .NET 6.0.3
Process Architecture: X64
OS Description: Microsoft Windows 10.0.19043


[DEBUG:0@1.820] global D:\bb\cv_x64_cuda\build\opencv\modules\core\src\parallel\registry_parallel.impl.hpp (69) cv::parallel::ParallelBackendRegistry::ParallelBackendRegistry core(parallel): Builtin backends(3): ONETBB(1000); TBB(990); OPENMP(980)
[DEBUG:0@1.820] global D:\bb\cv_x64_cuda\build\opencv\modules\core\src\parallel\registry_parallel.impl.hpp (94) cv::parallel::ParallelBackendRegistry::ParallelBackendRegistry core(parallel): Available backends(3): ONETBB(1000); TBB(990); OPENMP(980)
[ INFO:0@1.820] global D:\bb\cv_x64_cuda\build\opencv\modules\core\src\parallel\registry_parallel.impl.hpp (96) cv::parallel::ParallelBackendRegistry::ParallelBackendRegistry core(parallel): Enabled backends(3, sorted by priority): ONETBB(1000); TBB(990); OPENMP(980)
Available Parallel backends:
ONETBB
TBB
OPENMP

DepthAI module available.

[ INFO:1@2.151] global D:\bb\cv_x64_cuda\build\opencv\modules\core\src\utils\plugin_loader.impl.hpp (74) cv::plugin::impl::DynamicLib::libraryRelease unload D:\EmguCV\Emgu.CV.Example\BuildInfo.NetCore.Console\bin\Debug\net6.0\runtimes\win-x64\native\opencv_videoio_ffmpeg455_64.dll

D:\EmguCV\Emgu.CV.Example\BuildInfo.NetCore.Console\bin\Debug\net6.0\BuildInfo.NetCore.Console.exe (process 13120) exited with code 0.
To automatically close the console when debugging stops, enable Tools->Options->Debugging->Automatically close the console when debugging stops.
Press any key to close this window . . .

**Comment by PxPete** [comment]
Interestingly if I run the XamarinForms.WPF sample and try the MaskRcnn using CUDA I also get an error...

![image](https://user-images.githubusercontent.com/102365959/160158958-d4d3d0b9-2f86-4d4f-83b0-a7427bf08bca.png)

**Reply by emgucv** [reply]
Your system have a NVS 310 GPU, which is based on CUDA compute 2.1
```
Platform 1: NVIDIA CUDA
Device 0 (Default): NVS 310 1.1 (Gpu):Version - OpenCL 1.1 CUDA; Global memory - 1073741824; Local memory - 49152; Max image size - 16384x16384; DoubleFpConfig: Denorm, InfNan, RoundToNearest, RoundToZero, RoundToInf, Fma
```

Our release is build to require at least CUDA compute 5.6:
`NVIDIA GPU arch: 52 60 61 70 75 80 86`

You will need a more recent version of CUDA device from Nvidia. For a list of compatible GPUs please check the following:
https://en.wikipedia.org/wiki/CUDA#GPUs_supported

**Reply by PxPete** [reply]
Thanks for getting back to me again. The application we have in mind is very cost sensitive hence the older standard of card. I do have a newer card waiting for me to collect from the office which I think (looking at the document in the link) should support the required version - fingers crossed.
